src/python/pants/binaries/binary_tool.py

```diff
@@ -230,3 +297,113 @@ def tar_xz_extractor(self):
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class VersionDigestMapping:
+  """Parse the --version-digest-mapping option back into a dictionary."""
+  version_digest_mapping: Tuple[Tuple[str, Tuple[str, str, int]], ...]
+
+  def __init__(self, version_digest_mapping: Dict[str, List[Union[str, int]]]) -> None:
+    self.version_digest_mapping = tuple(
+      (platform_constraint, tuple(data)) # type: ignore[misc]
+      for platform_constraint, data in version_digest_mapping.items()
+    )
+
+  @memoized_property
+  def _deserialized_mapping(
+      self,
+  ) -> Dict[PlatformConstraint, ToolForPlatform]:
+    deserialized: Dict[PlatformConstraint, ToolForPlatform] = {}
+    for platform_constraint, (version, fingerprint, size_bytes) in self.version_digest_mapping:
+      deserialized[PlatformConstraint(platform_constraint)] = ToolForPlatform(
+        version=ToolVersion(version),
+        digest=Digest(fingerprint, size_bytes),
+      )
+    return deserialized
+
+  def get(self, platform_constraint: PlatformConstraint) -> ToolForPlatform:
+    return self._deserialized_mapping[platform_constraint]
+
+
+@dataclass(frozen=True)
+class BinaryToolUrlSet:
+  tool_for_platform: ToolForPlatform
+  host_platform: HostPlatform
+  url_generator: BinaryToolUrlGenerator
+
+  def get_urls(self) -> List[str]:
+    return self.url_generator.generate_urls(
+      version=self.tool_for_platform.version.version,
+      host_platform=self.host_platform if self.host_platform != HostPlatform.empty else None)
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class BinaryToolFetchRequest:
+  tool: BinaryToolBase
+  platform_constraint: PlatformConstraint
+
+  def __init__(
+      self,
+      tool: BinaryToolBase,
+      platform_constraint: Optional[PlatformConstraint] = None,
+  ) -> None:
+    self.tool = tool
+    if platform_constraint is None:
+      if tool.platform_dependent:
+        platform_constraint = PlatformConstraint.local_platform
+      else:
+        platform_constraint = PlatformConstraint.none
+
+    self.platform_constraint = platform_constraint
+
+
+@rule
+async def get_binary_tool_urls(
+    req: BinaryToolFetchRequest,
+    binary_util: BinaryUtil,
+) -> BinaryToolUrlSet:
+  tool = req.tool
+  platform_constraint = req.platform_constraint
+
+  mapping = VersionDigestMapping(tool.get_options().version_digest_mapping)
+  tool_for_platform = mapping.get(platform_constraint)
+
+  version = tool_for_platform.version.version
+  url_generator = binary_util.get_url_generator(tool.make_binary_request(version))
+
+  host_platform = await Get[HostPlatform](PlatformConstraint, platform_constraint)
+
+  return BinaryToolUrlSet(
+    tool_for_platform=tool_for_platform,
+    host_platform=host_platform,
+    url_generator=url_generator,
+  )
+
+
+@rule
+async def fetch_binary_tool(req: BinaryToolFetchRequest) -> Snapshot:
+  url_set = await Get[BinaryToolUrlSet](BinaryToolFetchRequest, req)
```

Comment:
I think that `BinaryToolUrlSet` could go into your parameter list rather than yielding here... alternatively, could inline that rule I think.